File: tests/c/flogadd2.c
/* Source File: flogadd1.c */

#ifdef __CINT__
#include "cint_types.h"
#endif

#ifndef __CINT__
#include <stdio.h>
#include <stdlib.h>
#include <interface.h>
#define MAXRECORDS 1000000
#endif

/*#define MAX_SPLIT_LENGTH 256*/

int main (int argc, char *argv[])
{
  FILE *fp;
  int status, i;
  char tbl1[TABLE_NAME_WIDTH+1]; 
  char tbl2[TABLE_NAME_WIDTH+1]; 
  char ssn[10];
  char *field;
  double t1, t2;
  int numLoops = 0;
  int numLicenses;
  ListHeader *lh;
  Link *lnk;
  int counter, modulus, mod_inc;

  /*
   * First, check the command-line arguments.
   */
  if (argc == 1) {
    printf ("\n\nUsage: %s numloops\n", argv[0]);
    printf ("  Where numloops is the number of times to iterate\n");
    return 1;
  }
  if (argc == 2) {
    numLoops = atoi (argv[1]);
    if (numLoops > MAXRECORDS || 
	numLoops < 0) {
      printf ("\n\n***Error: invalid value for numLoops: %d\n", numLoops);
      return _ERROR_;
    }
  }
  /*
   * Now, create the list for SSNs.
   */
  lh = initList (UNSORTED, 0);
  if (0 == lh) {
    printf ("\n\n***Error: could not initialize list: %s\n",
	ListErrorString[ListError]);
    return _ERROR_;
  }

  /*
   * Next, remove old .db & co. files.
   */
  if (fexists ("flogadd2_1.db")) {
    fileRemove ("flogadd2_1.db");
    fileRemove ("flogadd2_1.db.LCK");
    fileRemove ("f2_1_ssnidx.inx");
    fileRemove ("f2_1_ssnidx.idx");
    fileRemove ("f2_1_lfmidx.inx");
    fileRemove ("f2_1_lfmidx.idx");
  }
  if (fexists ("flogadd2_2.db")) {
    fileRemove ("flogadd2_2.db");
    fileRemove ("flogadd2_2.db.LCK");
    fileRemove ("f2_2_ssnidx.inx");
    fileRemove ("f2_2_ssnidx.idx");
  }
  fp = fopen ("flogadd2.df", "wb");
  if (0 == fp) {
    printf ("\n\n***Error: could not open flogadd2.df to write\n");
    return _ERROR_;
  }
  fprintf (fp, 
      "// flogadd2.df\n\n"
      "create table \"flogadd2_1.db\"\n"
      "  info \"Table for flogadd2.c\"\n"
      "{\n"
      "  \"LName\" char (65);\n"
      "  \"SSNumber\" char (9);\n"
      "  \"FName\" char (35);\n"
      "  \"NumberLicenses\" number (2:0);\n"
      "} indexed {\n"
      "  idx \"f2_1_ssnidx\" 128:case:unique \"SSNumber\";\n"
      "  midx \"f2_1_lfmidx\" 512:case \"LName\", \"FName\";\n"
      "};\n\n"
      "create table \"flogadd2_2.db\"\n"
      "  info \"Table for flogadd2.c\"\n"
      "{\n"
      "  \"SSNumber\" char (9);\n"
      "  \"License\" char (7);\n"
      "} indexed {\n"
      "  idx \"f2_2_ssnidx\" 128:case:dup \"SSNumber\";\n"
      "};\n\n"
      );
  fclose (fp);

  status = dbcreate ("flogadd2.df");
  if (status == _ERROR_) {
    printf ("\n\n***Error: Could not dbcreate flogadd2.df: %s\n", cdberror);
    return _ERROR_;
  }

  field = (char *)dbopen ("flogadd2_1.db");
  if (0 == field) {
    printf ("\n\n***Error: Could not dbopen flogadd2_1.db: %s\n", cdberror);
    return _ERROR_;
  }
  memset (tbl1, 0, TABLE_NAME_WIDTH+1);
  strcpy (tbl1, field);

  field = (char *)dbopen ("flogadd2_2.db");
  if (0 == field) {
    printf ("\n\n***Error: Could not dbopen flogadd2_2.db: %s\n", cdberror);
    return _ERROR_;
  }
  memset (tbl2, 0, TABLE_NAME_WIDTH+1);
  strcpy (tbl2, field);

  status = fileRemove ("flogadd2.df");
  if (status == _ERROR_) {
    printf ("\n\n***Warning: could not remove flogadd2.df\n");
  }

  /*
   * OK, populate the table with data generated from the dbtest* commands.
   */

  t1 = dbtime ();
  counter = 0;
  modulus = 10;
  mod_inc = 1;
  while (counter < numLoops) {
    int RAND;
    counter++;
    if (counter%1000 == 0) {
      printf (".");
      fflush (stdout);
    }
    field = dbtestupperstring (0, 65);
    status = dbsetchar (tbl1, "LName", field);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetchar (%s, %s, %s): %s\n",
	  tbl1, "LName", field, cdberror);
      dbexit ();
      return _ERROR_;
    }
    field = dbtestnumber (9, 0);
    status = dbsetchar (tbl1, "SSNumber", field);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetchar (%s, %s, %s): %s\n",
	  tbl1, "SSNumber", field, cdberror);
      dbexit ();
      return _ERROR_;
    }
    strcpy (ssn, field);
    field = dbtestupperstring (0, 35);
    status = dbsetchar (tbl1, "FName", field);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetchar (%s, %s, %s): %s\n",
	  tbl1, "FName", field, cdberror);
      dbexit ();
      return _ERROR_;
    }
    /* generate a psuedo random value */
    t2 = dbtime();
    RAND = (long)t2*1000000L;
    RAND %= modulus;
    if (RAND <= 0)
      RAND = 10;
    modulus += mod_inc;
    if (modulus > 40)
      mod_inc = -1;
    if (modulus < 10)
      mod_inc = 1;
    status = dbsetnum (tbl1, "NumberLicenses", (int)RAND);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsetnum (%s, %s, %d): %s\n",
 	 tbl1, "NumberLicenses", RAND, cdberror);
      dbexit();
      return _ERROR_;
    }
    /* Add'em */
    status = dbadd (tbl1);
    if (status == _ERROR_) {
      if (!strcmp (cdberror, 
	    "adding record: unique index constraint violated")) {
	counter--;
	continue;
      }
      printf ("\n\n***Error: dbadd(%s): %s\n", tbl1, cdberror);
      dbexit ();
      return _ERROR_;
    }
    /* Now, add RAND items to the secondary table */
    for (i = 0; i < RAND; i++) {
      status = dbsetchar (tbl2, "SSNumber", ssn);
      if (status == _ERROR_) {
        printf ("\n\n***Error: dbsetchar(%s, %s, %s): %s\n",
 	   tbl2, "SSNumber", ssn, cdberror);
        dbexit ();
        return _ERROR_;
      }
      field = dbtestnumstring (7, 0);
      status = dbsetchar (tbl2, "License", field);
      if (status == _ERROR_) {
        printf ("\n\n***Error: dbsetchar(%s, %s, %s): %s\n",
	    tbl2, "License", field, cdberror);
	dbexit();
	return _ERROR_;
      }
      status = dbadd (tbl2);
      if (status == _ERROR_) {
        printf ("\n\n***Error: dbadd(%s): %s\n", tbl2, cdberror);
        dbexit();
        return _ERROR_;
      }
    }
    lnk = malloc (sizeof(Link));
    if (0 == lnk) {
      printf ("\n\n***Error: Critical memory error allocating Link\n");
      dbexit();
      return _ERROR_;
    }
    field = malloc (10);
    if (0 == field) {
      printf ("\n\n***Error: Critical memory error allocating for ssn\n");
      dbexit ();
      return _ERROR_;
    }
    strcpy (field, ssn);
    check_pointer (field);
    lnk->data = field;
    check_pointer (lnk);
    insertLink (lh, lnk);
  }
  t2 = dbtime();
  printf ("\n\nFinished adding...\n%d items in %s, %d items in %s\n",
      dbnumrecs (tbl1), tbl1, dbnumrecs(tbl2), tbl2);
  printf ("%d items added total\n", dbnumrecs(tbl1)+dbnumrecs(tbl2));
  printf ("Done in %6.4f seconds\n\n", t2-t1);

  /*
   * Now, traverse the table and make sure that the data is consistent with
   * what was entered.
   */
  counter = 0;
  dbcurrent (tbl1, "f2_1_ssnidx");
  dbcurrent (tbl2, "f2_2_ssnidx");
  status = dbiseof (tbl1);
  if (status == 1) {
    printf ("\n\n***Error: There are no records in %s on traversal\n", tbl1);
    dbexit ();
    return _ERROR_;
  }
  for (lnk = lh->head->next; lnk != lh->tail; lnk = lnk->next) {
    counter++;
    status = dbsearchexact (tbl1, (char*)lnk->data);
    if (status == 0) {
      printf ("\n\n***Warning: couldn't find \"%s\" in %s: item %d\n",
	  (char*)lnk->data, tbl1, counter);
      continue;
    }
    status = dbretrieve (tbl1);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbretrieve(%s): %s\n",
	  tbl1, cdberror);
      dbexit();
      return _ERROR_;
    }
    if (status == 0) {
      /* skip deleted */
      continue;
    }
    field = dbshow (tbl1, "NumberLicenses");
    if (0 == field) {
      printf ("\n\n***Error: dbshow (%s, %s): %s\n", tbl1, "NumberLicenses",
	  cdberror);
      dbexit();
      return _ERROR_;
    }
    numLicenses = atoi (field);
    if (numLicenses < 0 || numLicenses > 100) {
      printf ("\n\n***Error: invalid value for numLicenses: %d\n", numLicenses);
      dbexit ();
      return _ERROR_;
    }
    status = dbsearchexact (tbl2, (char*)lnk->data);
    if (status == _ERROR_) {
      printf ("\n\n***Error: dbsearchexact (%s, %s): %s\n",
	  tbl2, (char*)lnk->data, cdberror);
      dbexit ();
      return _ERROR_;
    }
    if (status == 0) {
      printf ("\n\n***Warning: could not find %s in %s\n", (char*)lnk->data,
	  tbl2);
      continue;
    }
    for (i = 0; i < numLicenses;i++) {
      status = dbretrieve(tbl2);
      if (status == _ERROR_) {
	printf ("\n\n***Error: dbretrieve(%s): %s\n", tbl2, cdberror);
	dbexit ();
	return _ERROR_;
      }
      if (status == 0) {
	/* shouldn't be any deleted records, but what the heck. */
	dbnextindex (tbl2);
	continue;
      }
      field = dbshow (tbl2, "SSNumber");
      if (strcmp ((char*)lnk->data, field)) {
	printf ("\n\n***Warning: ssn from %s (%s) != ssn from %s (%s)\n",
	    tbl1, (char*)lnk->data, tbl2, field);
      }
      dbnextindex (tbl2);
    }
  }
  t2 = dbtime();

  /*
   * Clean up and hit it.
   */
  lnk = removeLink (lh);
  while (lnk != 0) {
    free (lnk->data);
    free (lnk);
    lnk = removeLink (lh);
  }
  delList (lh);
  printf ("\n\nVerified the table data in %6.4f seconds\n", t2-t1);
  dbexit ();
  print_block_list();
  return 0;
}

File: tests/python/flogadd2.py
#!/usr/bin/python2.5

# Load the python module.
import sys
import os
sys.path[0] = ''
os.chdir ("../../lib")
import cdbpy
os.chdir ("../tests/python")

# First, get rid of the old stuff, if it exists.
if (os.access ("flogadd2_1.db", os.F_OK)) :
  os.remove ("flogadd2_1.db")
if (os.access ("flogadd2_1.db.LCK", os.F_OK)) :
  os.remove ("flogadd2_1.db.LCK")
if (os.access ("f2_1_ssnidx.inx", os.F_OK)) :
  os.remove ("f2_1_ssnidx.inx")
if (os.access ("f2_1_ssnidx.idx", os.F_OK)) :
  os.remove ("f2_1_ssnidx.idx")
if (os.access ("f2_1_lfmidx.inx", os.F_OK)) :
  os.remove ("f2_1_lfmidx.inx")
if (os.access ("f2_1_lfmidx.idx", os.F_OK)) :
  os.remove ("f2_1_lfmidx.idx")

if (os.access ("flogadd2_2.db", os.F_OK)) :
  os.remove ("flogadd2_2.db")
if (os.access ("flogadd2_2.db.LCK", os.F_OK)) :
  os.remove ("flogadd2_2.db.LCK")
if (os.access ("f2_2_ssnidx.inx", os.F_OK)) :
  os.remove ("f2_2_ssnidx.inx")
if (os.access ("f2_2_ssnidx.idx", os.F_OK)) :
  os.remove ("f2_2_ssnidx.idx")

fadd2 = open ("flogadd2.df", "w", 0644)
file.write (fadd2, '// flogadd2.df\n')
file.write (fadd2, 'create table "flogadd2_1.db"\n')
file.write (fadd2, '  info "Table for flogadd2.py"\n')
file.write (fadd2, '{\n')
file.write (fadd2, '  "LName" char (65);\n')
file.write (fadd2, '  "SSNumber" char (9);\n')
file.write (fadd2, '  "FName" char (35);\n')
file.write (fadd2, '  "NumberLicenses" number (2:0);\n')
file.write (fadd2, '} indexed {\n')
file.write (fadd2, '  idx "f2_1_ssnidx" 256:case:unique "SSNumber";\n')
file.write (fadd2, '  midx "f2_1_lfmidx" 512:case "LName", "FName";\n')
file.write (fadd2, '};\n\n')
file.write (fadd2, 'create table "flogadd2_2.db"\n')
file.write (fadd2, '  info "Table for flogadd2.py"\n')
file.write (fadd2, '{\n')
file.write (fadd2, '  "SSNumber" char (9);\n')
file.write (fadd2, '  "License" char (7);\n')
file.write (fadd2, '} indexed {\n')
file.write (fadd2, '  idx "f2_2_ssnidx" 256:case:dup "SSNumber";\n')
file.write (fadd2, '};\n\n')
file.close (fadd2)

status = cdbpy.dbcreate ("flogadd2.df")
if (status != 0) :
  print "%s %s" % ("\n\n***Error dbcreate flogadd2.df: ", cdbpy.dbgeterror ())
  sys.exit (1)

# Fire at will.
table = cdbpy.dbopen ("flogadd2_1.db")
if (table == '') :
  print "\n\n***Error: dbopen flogadd2_1.db: %s" % (cdbpy.dbgeterror())
  sys.exit (1)
tbl1 = cdbpy.dbopen ("flogadd2_2.db")
if (table == '') :
  print "\n\n***Error: dbopen flogadd2_2.db: %s" % (cdbpy.dbgeterror())
  sys.exit (1)

os.remove ("flogadd2.df")

counter = 0
modulus = 10
mod_inc = 1
ssnList = []

t1 = 0.0
t1 = cdbpy.dbtime ()

import random
random.seed (None)

execfile ("fa_inc.py")
while (counter < numLoop) :
  counter += 1
  if (counter % 1000 == 0) :
    sys.stdout.write ('.')
    sys.stdout.flush()
  LName = cdbpy.dbtestupperstring (0, 65)
  if (LName == '') :
    print "\n\n***Error: dbtestupperstring 0 65: %s\n" % \
        (cdbpy.dbgeterror ())
    cdbpy.dbexit ()
    sys.exit (1)
  status = cdbpy.dbsetchar (table, "LName", LName)
  if (status == -1) :
    print "\n\n***Error: dbsetchar %s LName %s:%s" % \
        (table, LName, cdbpy.dbgeterror())
    cdbpy.dbexit ()
    sys.exit (1)
  SSNumber = cdbpy.dbtestnumber (9, 0)
  if (SSNumber == '') :
    print "\n\n***Error: dbtestnumber 9 0: %s\n" % \
        (cdbpy.dbgeterror ())
    cdbpy.dbexit ()
    sys.exit (1)
  status = cdbpy.dbsetchar (table, "SSNumber", SSNumber)
  if (status == -1) :
    print "\n\n***Error: dbsetchar %s SSNumber %s:%s" % \
        (table, SSNumber, cdbpy.dbgeterror())
    cdbpy.dbexit ()
    sys.exit (1)
  FName = cdbpy.dbtestupperstring (0, 35)
  if (FName == '') :
    print "\n\n***Error: dbtestupperstring 0 35: %s\n" % \
        (cdbpy.dbgeterror ())
    cdbpy.dbexit ()
    sys.exit (1)
  status = cdbpy.dbsetchar (table, "FName", FName)
  if (status == -1) :
    print "\n\n***Error: dbsetchar %s FName %s:%s" % \
        (table, FName, cdbpy.dbgeterror())
    cdbpy.dbexit ()
    sys.exit (1)
  RAND = random.randint (0, modulus)
  if (RAND == 0) : RAND = 10
  modulus += mod_inc
  if (modulus > 40) : mod_inc = -1
  if (modulus < 10) : mod_inc = 1
  status = cdbpy.dbsetnum (table, "NumberLicenses", RAND)
  if (status == -1) :
    print "\n\n***Error: dbsetnum %s NumberLicenses %s:%s" % \
        (table, RAND, cdbpy.dbgeterror())
    cdbpy.dbexit ()
    sys.exit (1)
  status = cdbpy.dbadd (table)
  if (status == -1) :
    terr = cdbpy.dbgeterror()
    status = cmp (terr, "adding record: unique index constraint violated")
    if not status :
      counter -= 1
      continue
    print "\n\n***Error: dbadd %s:%s\n" % (table, cdbpy.dbgeterror())
    cdbpy.dbexit ()
    sys.exit (1)
  newcount = 0
  while (newcount < RAND) :
    newcount += 1
    status = cdbpy.dbsetchar (tbl1, "SSNumber", SSNumber)
    if (status == -1) :
      print "\n\n***Error: dbsetchar %s SSNumber %s:%s" % \
          (tbl1, SSNumber, cdbpy.dbgeterror())
      cdbpy.dbexit ()
      sys.exit (1)
    license = cdbpy.dbtestnumstring (7, 0)
    if (license == '') :
      print "\n\n***Error: dbtestnumstring (7, 0): %s\n" % \
          (cdbpy.dbgeterror())
      cdbpy.dbexit()
      sys.exit (1)
    status = cdbpy.dbsetchar (tbl1, "License", license)
    if (status == -1) :
      print "\n\n***Error: dbsetchar %s License %s:%s" % \
          (tbl1, license, cdbpy.dbgeterror())
      cdbpy.dbexit ()
      sys.exit (1)
    status = cdbpy.dbadd (tbl1)
  ssnList.append (SSNumber)

cdbpy.dbclose (table)
cdbpy.dbclose (tbl1)

table = cdbpy.dbopen ("flogadd2_1.db")
if (table == '') :
  print "Opening flogadd2_1.db the second time: %s" % (cdbpy.dbgeterror())
  sys.exit (1)
tbl1 = cdbpy.dbopen ("flogadd2_2.db")
if (tbl1 == '') :
  print "Opening flogadd2_2.db the second time: %s" % (cdbpy.dbgeterror())
  cdbpy.dbclose (table)
  sys.exit (1)

t2 = 0.0
t2 = cdbpy.dbtime ()

num_table = cdbpy.dbnumrecs (table)
num_tbl1 = cdbpy.dbnumrecs (tbl1)

print "\nFinished adding...\n"
print "%d items in %s and %d items in %s\n" % \
    (num_table, table, num_tbl1, tbl1)
print "Done in %f seconds" % (t2-t1)

counter = 0
cdbpy.dbcurrent (table, "f2_1_ssnidx")
cdbpy.dbcurrent (tbl1, "f2_2_ssnidx")
status = cdbpy.dbiseof (table)
if (status == 1) :
  print "There are no records in %s\n" % (table)
  cdbpy.dbexit ()
  sys.exit (1)

t1 = 0.0
t1 = cdbpy.dbtime ()

for ssn in ssnList :
  counter += 1
  status = cdbpy.dbsearchexact (table, ssn)
  if (status == 0) :
    print "Could not retrieve from %s: %s\n" % (table, cdbpy.dbgeterror())
    continue
  status = cdbpy.dbretrieve (table)
  if (status == 0) : continue  # deleted record
  if (status == -1) :
    print "\n\n***Error: dbretrieve %s: %s\n" % \
        (table, cdbpy.dbgeterror())
    cdbpy.dbexit ()
    sys.exit (1)
  numlic = cdbpy.dbshow (table, "NumberLicenses")
  if (numlic == '') : continue
  numLicenses = int (numlic)
  status = cdbpy.dbsearchexact (tbl1, ssn)
  if (status == 0) :
    print "***Warning: Couldn't find %s in %s\n" % \
        (ssn, tbl1)
    continue
  i = 0
  while (i < numLicenses) :
    i += 1
    status = cdbpy.dbretrieve (tbl1)
    if (status == 0) : cdbpy.dbnextindex (tbl1)
    if (status == -1) :
      print "\n\n***Error: dbretrieve %s: %s\n" % \
          (tbl1, cdbpy.dbgeterror())
      cdbpy.dbexit ()
      sys.exit (1)
    this_ssn = cdbpy.dbshow (tbl1, "SSNumber")
    if (this_ssn == '') : continue
    if (this_ssn != ssn) :
      print "***Warning: ssn from %s (%s) != ssn from %s (%s)\n" % \
          (table, ssn, tbl1, this_ssn)
    cdbpy.dbnextindex (tbl1)

cdbpy.dbexit ()
if (os.access ("fa_inc.py", os.F_OK)) :
  os.remove ("fa_inc.py")

t2 = 0.0
t2 = cdbpy.dbtime ()

print "\n\nVerified the table data in %f seconds" % (t2-t1)
sys.exit (0)
